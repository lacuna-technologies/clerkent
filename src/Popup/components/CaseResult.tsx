import useClipboard from 'Popup/hooks/useClipboard'
import type { FunctionComponent } from 'preact'
import { useCallback } from 'preact/hooks'
import Helpers from 'utils/Helpers'
import ResultLink from './ResultLink'

interface Props {
  case: Law.Case,
  downloadPDF: downloadPDFType
}

const CaseResult: FunctionComponent<Props> = ({
  case: currentCase,
  downloadPDF,
}) => {
  const {
    citation,
    name,
    links,
    database,
  } = currentCase
  const summaryURL = Helpers.getSummaryLink(links)?.url
  const judgmentLink = Helpers.getJudgmentLink(links)
  const opinionLink = Helpers.getOpinionLink(links)
  const orderLink = Helpers.getOrderLink(links)

  const caseNameClass = summaryURL ? `text-blue-700 border-0 bg-none outline-none p-0 underline cursor-pointer select-text hover:text-blue-900 hover:underline` : ``
  const {
    permissionGranted,
    promptGrant,
  } = useClipboard()
  const onClickCitation = useCallback(async () => {
    if(!permissionGranted){
      await promptGrant()
    }
    await navigator.clipboard.writeText(citation)
  }, [permissionGranted, promptGrant, citation])

  return (
    <div className="w-full outline-none">
      <div className="flex flex-row justify-start items-center mb-0.5">
        {
          database && (
            <a className="text-xs" href={database.url} target="_blank" rel="noreferrer">
              {database.name}
            </a>
          )
        }
      </div>
      <a
        className={`align-left text-lg ${caseNameClass}`}
        target="_blank"
        href={summaryURL}
        rel="noreferrer"
        {...(summaryURL ? {title: `Summary`} : {})}
      >
        {name}
      </a>
      <div className="flex flex-row mt-0.5 gap-4">
        <span
          className="cursor-pointer select-all"
          onClick={onClickCitation}
          title="Click to copy citation"
        >
          {citation}
        </span>
        <ResultLink
          link={judgmentLink}
          onDownloadPDF={downloadPDF({ doctype: `Judgment`, law: currentCase })}
        />
        <ResultLink
          link={opinionLink}
          onDownloadPDF={downloadPDF({ doctype: `Opinion`, law: currentCase })}
        />
        <ResultLink
          link={orderLink}
          onDownloadPDF={downloadPDF({ doctype: `Order`, law: currentCase })}
        />
      </div>
    </div>
  )
}

export default CaseResult