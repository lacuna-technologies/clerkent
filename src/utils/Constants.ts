import type Law from '../types/Law'

const JURISDICTIONS = {
  AU: {
    emoji: `ðŸ‡¦ðŸ‡º`,
    id: `AU` as Law.JursidictionCode,
    name: `Australia`,
  },
  CA: {
    emoji: `ðŸ‡¨ðŸ‡¦`,
    id: `CA` as Law.JursidictionCode,
    name: `Canada`,
  },
  EU: {
    emoji: `ðŸ‡ªðŸ‡º`,
    id: `EU` as Law.JursidictionCode,
    name: `European Union`,
  },
  HK: {
    emoji: `ðŸ‡­ðŸ‡°`,
    id: `HK` as Law.JursidictionCode,
    name: `Hong Kong`,
  },
  MY: {
    emoji: `ðŸ‡²ðŸ‡¾`,
    id: `MY` as Law.JursidictionCode,
    name: `Malaysia`,
  },
  NZ: {
    emoji: `ðŸ‡³ðŸ‡¿`,
    id: `NZ` as Law.JursidictionCode,
    name: `New Zealand`,
  },
  SG: {
    emoji: `ðŸ‡¸ðŸ‡¬`,
    id: `SG` as Law.JursidictionCode,
    name: `Singapore`,
  },
  UK: {
    emoji: `ðŸ‡¬ðŸ‡§`,
    id: `UK` as Law.JursidictionCode,
    name: `United Kingdom`,
  },
}


const SG_DATABASES = {
  lawnet: {
    icon: ``,
    name: `Lawnet`,
    url: `https://www.lawnet.sg/lawnet/web/lawnet/home`,
  },
  lexis: {
    icon: ``,
    name: `Lexread`,
    url: `https://www.lexread.lexisnexis.com/`,
  },
  sc: {
    icon: ``,
    name: `Supreme Court`,
    url: `https://www.supremecourt.gov.sg/news/supreme-court-judgments`,
  },
  slw: {
    icon: ``,
    name: `Singapore Law Watch`,
    url: `https://www.singaporelawwatch.sg/Judgments`,
  },
  sso: {
    icon: ``,
    name: `Singapore Statutes Online`,
    url: `https://sso.agc.gov.sg/`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw Asia`,
    url: `https://www.westlawasia.com/singapore/`,
  },
}

const UK_DATABASES = {
  bailii: {
    icon: ``,
    name: `BAILII`,
    url: `https://bailii.org/`,
  },
  legislation: {
    icon: ``,
    name: `legislation.gov.uk`,
    url: `https://www.legislation.gov.uk/`,
  },
  lexis: {
    icon: ``,
    name: `LexisNexis`,
    url: `http://www.lexisnexis.com/uk/legal`,
  },
  westlaw: {
    icon: ``,
    name: `Westlaw`,
    url: `https://uk.westlaw.com/Browse/Home/WestlawUK`,
  },
}

const EU_DATABASES = {
  curia: {
    icon: ``,
    name: `CURIA`,
    url: `https://curia.europa.eu/juris/recherche.jsf?language=en`,
  },
  epo: {
    icon: ``,
    name: `EPO`,
    url: `https://www.epo.org/law-practice/case-law-appeals`,
  },
}

const HK_DATABASES = {
  hkliihk: {
    icon: ``,
    name: `HKLII.hk`,
    url: `https://www.hklii.hk/`,
  },
  hkliiorg: {
    icon: ``,
    name: `HKLII.org`,
    url: `https://www.hklii.org/eng/`,
  },
}

const CA_DATABASES = {
  canlii: {
    icon: ``,
    name: `CanLII`,
    url: `https://www.canlii.org/en`,
  },
}

const AU_DATABASES = {
  austlii: {
    icon: ``,
    name: `AustLII`,
    url: `https://www8.austlii.edu.au`,
  },
}

const MISC_DATABASES = {
  commonlii: {
    icon: ``,
    name: `CommonLII`,
    url: `http://www.commonlii.org`,
  },
}

const dedupeObjects = (inputObject: Record<string, Record<string, string>>, jurisdictionCode: Law.JursidictionCode) =>
  Object.entries(inputObject)
  .reduce((accumulator, [id, object]: [string, Record<string, string>]) => ({
    ...accumulator,
    [`${jurisdictionCode}_${id}`]: object,
  }), {})

const DATABASES: Record<string, Law.Database> = {
  ...dedupeObjects(SG_DATABASES, `SG`),
  ...dedupeObjects(UK_DATABASES, `UK`),
  ...dedupeObjects(EU_DATABASES, `EU`),
  ...dedupeObjects(HK_DATABASES, `HK`),
  ...dedupeObjects(CA_DATABASES, `CA`),
  ...dedupeObjects(AU_DATABASES, `AU`),
  ...MISC_DATABASES,
}

const SG_COURTS = {
  sgca: {
    id: `SGCA`,
    name: `Court of Appeal`,
  },
  sghc: {
    id: `SGHC`,
    name: `High Court`,
  },
}

const EU_COURTS = {
  cjeu: {
    id: `CJEU`,
    name: `Court of Justice of the European Union`,
  },
  epo: {
    id: `EPO`,
    name: `European Patent Office`,
  },
}

const COURTS: Record<string, Record<string, string>> = {
  ...dedupeObjects(SG_COURTS, `SG`),
  ...dedupeObjects(EU_COURTS, `EU`),
}

const Constants = {
  COURTS,
  DATABASES,
  JURISDICTIONS,
}

export default Constants