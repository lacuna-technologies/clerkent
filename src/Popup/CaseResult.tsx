import React from 'react'
import Constants from '../utils/Constants'

const CaseResult = ({
  citation,
  name,
  link,
  pdf,
  jurisdiction,
  database,
  downloadPDF,
}) => {
  return (
    <div className="result">
      <div className="details">
        <div className="left">
          <span className="jurisdiction">{Constants.JURISDICTIONS[jurisdiction]?.emoji}</span>
          { database && <span className="database">{database.name}</span> }
        </div>
        <div className="right">
          {
            pdf ? (
              <p className="links">
                <button className="pdf button" onClick={downloadPDF({ citation, name, pdf })}>PDF</button>
              </p>
            ) : null
          }
        </div>
      </div>
      <a
        className="case-name link"
        target="_blank"
        href={link}
        rel="noreferrer"
      >
        {name} {citation}
      </a>
    </div>
  )
}

export default CaseResult