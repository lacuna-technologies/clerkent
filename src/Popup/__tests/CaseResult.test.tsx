import React from 'react'
import renderer from 'react-test-renderer'
import Constants from '../../utils/Constants'
import CaseResult from '../CaseResult'

const mockCase: Law.Case = {
  citation: `[2006] EWCA Civ 145`,
  database: Constants.DATABASES.UK_bailii,
  jurisdiction: Constants.JURISDICTIONS.UK.id,
  links: [
    {
      doctype: `Judgment`,
      filetype: `HTML`,
      url: `https://www.bailii.org/ew/cases/EWCA/Civ/2006/145.html`,
    },
  ],
  name: `IDA v University of Southampton`,
  type: `case-name`,
}
const mockDownloadPDF = () => () => {}

describe(`CaseResult`, () => {
  it(`renders without error`, () => {
    const tree = renderer.create(
      <CaseResult
        downloadPDF={mockDownloadPDF}
        case={mockCase}
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})