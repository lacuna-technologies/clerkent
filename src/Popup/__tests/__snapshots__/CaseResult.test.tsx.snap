// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseResult renders without error 1`] = `
<div
  className="result"
>
  <div
    className="details"
  >
    <div
      className="left"
    >
      <span
        className="jurisdiction"
      >
        ðŸ‡¬ðŸ‡§
      </span>
      <span
        className="database"
      >
        BAILII
      </span>
    </div>
    <div
      className="right"
    />
  </div>
  <a
    className="case-name link"
    href="https://www.bailii.org/ew/cases/EWCA/Civ/2006/145.html"
    rel="noreferrer"
    target="_blank"
  >
    IDA v University of Southampton
     
    [2006] EWCA Civ 145
  </a>
</div>
`;
